examples/cpp/fill-grid-pp2hadron.cpp

```diff
@@ -0,0 +1,245 @@
+
+        // skip if kinematically forbidden
+        if (jacobian == 0.0 || z <= 0.0) {
+            continue;
+        }
+
+        // apply cuts on hadron kinematics
+        if ((pt_hadron < pt_min) || (pt_hadron > pt_max) || (fabs(y_hadron) > abs_y_max)) {
+            continue;
+        }
+
+        jacobian *= hbarc2 / calls;
+
+        // calculate the partonic cross-section
+        auto weight = jacobian * me_gg2qqbar(s, t, u);
+
+        double q2 = pt_hadron * pt_hadron;
+        std::size_t order = 0;
+        std::size_t channel = 0;
+
+        // define the tuple of kinematic variables `ntuples = (q2, x1, x2, z)`
+        std::vector<double> ntuples = {q2, x1, x2, z};
+
+        // Fill the grid using hadron `pT` as the observable
+        pineappl_grid_fill2(grid, order, pt_hadron, channel, ntuples.data(), weight);
+    }
+}
+
+int main() {
+    // ---
+    // Define the partonic channels and orders that will be filled into the grid
+
+    // specify the number of convolutions: 2 for initial-state PDFs + 1 for FFs
+    std::size_t nb_convolutions = 3;
+
+    // instantiate the channel object
+    auto* channels = pineappl_channels_new(nb_convolutions);
+
+    // specify the contributing channel(s) and the corresponding factor(s)
+    // for the process `gg -> qqbar` we need to sum over the light quarks
+    std::vector<int32_t> pids;
+    std::vector<double> factors;
+    for (int i = -3; i <= 3; ++i) {
+        if (i == 0) continue;
+        pids.insert(pids.end(), {21, 21, i});
+        factors.push_back(1.0);
+    }
+    pineappl_channels_add(channels, pids.size() / nb_convolutions, pids.data(),
+        factors.data());
+
+    // specify the perturbative orders that will be filled into the grid
+    // orders specifies the power of the tuple `orders = (αs, α, lR, lF, lD)`
+    // in this example, we only fill the LO QCD
+    std::vector<uint8_t> orders = {1, 0, 0, 0, 0};
+
+    // bin limits of the final-state hadron transverse momentum
+    std::vector<double> bins = {
+        5.0, 7.0, 10.0, 15.0, 20.0, 25.0, 30.0, 35.0, 40.0, 45.0, 50.0,
+        60.0, 70.0, 80.0, 90.0, 100.0
+    };
+
+    // ---
+    // Construct the objects that are needed to fill the grid
+
+    // choose the Evolution Basis to represent the grid
+    pineappl_pid_basis pid_basis = PINEAPPL_PID_BASIS_EVOL;
+    // define the types of hadrons and set them to be Unpolarised
+    pineappl_conv convs[] = {
+        {PINEAPPL_CONV_TYPE_UNPOL_PDF, 2212}, // proton
+        {PINEAPPL_CONV_TYPE_UNPOL_PDF, 2212}, // proton
+        {PINEAPPL_CONV_TYPE_UNPOL_FF, 211},   // pion
+    };
+
+    // define the kinematics object `kinematics = (μ, x1, x2, x)`
+    pineappl_kinematics scales = {PINEAPPL_KINEMATICS_SCALE, 0};
+    pineappl_kinematics x1 = {PINEAPPL_KINEMATICS_X, 0};
+    pineappl_kinematics x2 = {PINEAPPL_KINEMATICS_X, 1};
+    pineappl_kinematics z = {PINEAPPL_KINEMATICS_X, 2};
+    pineappl_kinematics kinematics[4] = {scales, x1, x2, z};
```

Comment:
If you initialize arrays like this, leave out the number of elements, which the compiler automatically infers. This is beneficial if you change it.